### PR TITLE
Fix getWindowPointerPosition on Wayland #177

### PR DESCRIPTION
**Pull request by simon-spinner**
Depending on the window currently focused the
coordinates returned may be wrong (sometimes even
negative). This happens mainly with popup windows.

Fixes https://github.com/eclipse-platform/eclipse.platform.swt/issues/177

### CODE REVIEW

**Comment by lalmeras** [comment]
Do you need someone to test this proposal with eclipse ? Is there something I can do to ease the validation of this PR ?

**Comment by lalmeras** [comment]
I try the proposed change and it fixes the issue pointed by #177 for me.

Process :
* use a Gnome + wayland session (Fedora 39 here)
* download eclipse 2024-03 for Java developers package
* vanilla eclipse reproduces #177 issue with GDK_BACKEND=wayland
  * Open declaration, implementation, ... menu closes before mouse cursor enters in menu
  * Javadoc over popup closes before mouse cursor move to popup
* retrieve eclipse.platform.swt, branch I20240227-1800 (match 2024-03 version)
* cherry-pick commit
* build bundles and binaries ; replace org.eclipse.swt-3.125.0-SNAPSHOT.jar and org.eclipse.swt.gtk.linux.x86_64_3.125.0.v20240227-1638.jar in plugins folder
* I cannot reproduce the issue :heart: 

Are there any other thing I can provide to ease the merge of this fix ?

File: bundles/org.eclipse.swt/Eclipse SWT/gtk/org/eclipse/swt/widgets/Display.java [hunk 1]
```diff
@@ -6203,7 +6203,27 @@ long getWindowPointerPosition(long window, int[] x, int[] y, int[] mask) {
...
+		if (windowAtPosition != 0 && windowAtPosition != window) {
+			int[] origin_x = new int[1], origin_y = new int[1];
+			GDK.gdk_window_get_origin(windowAtPosition, origin_x, origin_y);
+			x[0] += origin_x[0];
+			y[0] += origin_y[0];
+			GDK.gdk_window_get_origin(window, origin_x, origin_y);
+			x[0] -= origin_x[0];
+			y[0] -= origin_y[0];
+		}
+		return 0;
```
[the remark this answers is not in the document]
Comment (simon-spinner) [reply bundles/org.eclipse.swt/Eclipse SWT/gtk/org/eclipse/swt/widgets/Display.java]: According to the documentation gdk_window_get_device_position should return the same window pointer as gdk_device_get_window_at_position. However, the former always returns NULL on my computer whereas the latter returns a valid pointer. Changed the return statement accordingly.
Simple DnD looks ok to me in a quick test, but I am not familiar how it is supposed to work.

**Comment by lalmeras** [comment]
Some feedback about my tests performed by swapping eclipse swt with the patch provided by @simon-spinner (version `return 0`.

I use the patched version for day-to-day job and I encounter no issue with popup, and overall experience with eclipse is similar to Xorg + GDK_BACKEND=x11 version.

I will build a version with the updated patch (`return windowAtPosition`) so I can provide updated feedback.

**Comment by lalmeras** [comment]
Hi,

I use the last PR version for 1 month in a customized eclipse java build without any issue :heart:

I think it should be merged as it is a must-have for wayland usability.

Do you need any other feedback ?

**Comment by iloveeclipse** [comment]
@akurtakov : is it OK now for you?

**Comment by akurtakov** [comment]
Let's merge.

**Pull request merged**